File: Specialization_Data_Structures_Algorithms_UC_San_Diego/Course1_Algorithmic_Toolbox/week2_Introduction/fibonacci_last_digit.cpp
#include <iostream>

int get_fibonacci_last_digit_naive(int n) {
  if (n <= 1)
    return n;

  int previous = 0;
  int current = 1;

  for (int i = 0; i < n - 1; ++i) {
    int tmp_previous = previous;
    previous = current;
    current = tmp_previous + current;
  }

  return current;
}

int get_fibonacci_last_digit_fast(int n) {
  if (n <= 1)
    return n;

  int previous = 0;
  int current = 1;

  for (int i = 0; i < n - 1; ++i) {
    int tmp_previous = previous % 10;
    previous = current;
    current = (tmp_previous + current) % 10;
  }

  return current % 10;
}

int main() {
  int n;
  std::cin >> n;
  int c = get_fibonacci_last_digit_fast(n);
  std::cout << c << '\n';
  return 0;
}

File: Specialization_Data_Structures_Algorithms_UC_San_Diego/Course1_Algorithmic_Toolbox/week2_Introduction/fibonacci_last_digit/fibonacci_last_digit.py
# Uses python3
import sys

def get_fibonacci_last_digit_naive(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % 10

def get_fibonacci_last_digit_fast(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(n - 1):
        previous, current = current%10, (previous + current)%10

    return current % 10

if __name__ == '__main__':
    input = sys.stdin.read()
    n = int(input)
    print(get_fibonacci_last_digit_fast(n))
